src/uct/sm/knem/knem_pd.c

```diff
@@ -0,0 +1,210 @@
+/**
+* Copyright (C) Mellanox Technologies Ltd. 2001-2015.  ALL RIGHTS RESERVED.
+* Copyright (c) UT-Battelle, LLC. 2014-2015. ALL RIGHTS RESERVED.
+*
+* $COPYRIGHT$
+* $HEADER$
+*/
+
+#include "knem_pd.h"
+#include "knem_io.h"
+
+ucs_status_t uct_knem_pd_query(uct_pd_h pd, uct_pd_attr_t *pd_attr)
+{
+    pd_attr->rkey_packed_size  = sizeof(uct_knem_key_t);
+    pd_attr->cap.flags         = UCT_PD_FLAG_REG;
+    pd_attr->cap.max_alloc     = 0;
+    pd_attr->cap.max_reg       = ULONG_MAX;
+
+    memset(&pd_attr->local_cpus, 0xff, sizeof(pd_attr->local_cpus));
+    return UCS_OK;
+}
+
+static ucs_status_t uct_knem_query_pd_resources(uct_pd_resource_desc_t **resources_p,
+                                               unsigned *num_resources_p) {
+    int fd;
+    int rc;
+    struct knem_cmd_info info;
+
+    memset(&info, 0, sizeof(struct knem_cmd_info));
+
+    fd = open("/dev/knem", O_RDWR);
+    if (fd < 0) {
+        ucs_debug("Could not open the KNEM device file at /dev/knem: %m. Disabling knem resource");
+        *resources_p     = NULL;
+        *num_resources_p = 0;
+        return UCS_OK;
+    } else {
```

Comment:
to make it more "stylish", can avoid the "else" clause, since there is already "return OK" in the "then" case, just continue the function code after the if